src/ucs/datastruct/ptr_array.c

```diff
@@ -172,37 +180,126 @@ unsigned ucs_ptr_array_insert(ucs_ptr_array_t *ptr_array, void *value,
     }
 
     /* Get the first item on the free list */
-    elem_index = ptr_array->freelist;
-    ucs_assert(elem_index != UCS_PTR_ARRAY_SENTINEL);
-    elem = &ptr_array->start[elem_index];
+    element_index = ptr_array->freelist;
+    ucs_assert(element_index != UCS_PTR_ARRAY_SENTINEL);
 
-    /* Remove from free list */
+    elem = &ptr_array->start[element_index];
+
+    /* Remove from free list and populate */
     ptr_array->freelist = ucs_ptr_array_freelist_get_next(*elem);
+    *elem               = (uintptr_t)value;
 
-    /* Populate */
-    *placeholder_p = ucs_ptr_array_placeholder_get(*elem);
-    *elem = (uintptr_t)value;
-    return elem_index;
+    return element_index;
 }
 
-void ucs_ptr_array_remove(ucs_ptr_array_t *ptr_array, unsigned elem_index,
-                          uint32_t placeholder)
+void ucs_ptr_array_remove(ucs_ptr_array_t *ptr_array, unsigned element_index)
 {
-    ucs_ptr_array_elem_t *elem = &ptr_array->start[elem_index];
+    ucs_ptr_array_elem_t *elem = &ptr_array->start[element_index];
+    ucs_ptr_array_elem_t *next_elem;
+    uint32_t size_free_ahead;
+
+    ucs_assert_always(!ucs_ptr_array_is_free(ptr_array, element_index));
+
+    if (ucs_ptr_array_is_free(ptr_array, element_index + 1)) {
+        next_elem = &ptr_array->start[element_index + 1];
+        size_free_ahead = ucs_ptr_array_size_free_get(*next_elem);
+        size_free_ahead++;
```

Comment:
minor: i'd write this as `size_free_ahead = ucs_ptr_array_size_free_get(*next_elem) + 1`

Reply:
Fixed.